Add support for playing Bandcamp music from custom domains

jefetienne opened the pull request:
### Description of the changes

Previously unrecognizable to Red, this PR allows users to put in custom domain links of bandcamp pages and be able to play the track(s) listed on the page.

The audio cog will webscrape the HTML data of the unrecognized domain to find the original bandcamp url and use that url as the query, otherwise it will say it is not found.

Example:

From this page
```
https://jazzinbritain.co.uk/album/revisiting-tanglewood-63-the-early-tapes
```
we retrieve the bandcamp URL
```
https://jazzinbritain1.bandcamp.com/album/revisiting-tanglewood-63-the-early-tapes
```
and use that as the link to pull the music from.

### Have the changes in this PR been tested?

Without the unit tests I have added, it passes all of the tests. However, I am having problems trying to get my own test to run, but I think it might be more of a configuration issue on my end. I will write a comment below describing the issue.

jefetienne committed:
Play custom domain bandcamp music for audio cog
- Webscrape HTML of unrecognized domain to find the original bandcamp url and use that url as the query

jefetienne committed:
Add single track unit test

aikaterna commented:
I am of the opinion that this should be handled in the audio backend and not in our code, but I will wait to see what others have to say about this.

jefetienne commented:
This is the error I've been getting when I try to run tox with my tests. I've tried to model alias.py and test_alias.py to the best of my ability but I can't understand why this issue is happening.

```
_________________________________________________ ERROR collecting tests/cogs/test_audio.py __________________________________________________
.tox/py38/lib/python3.8/site-packages/redbot/core/data_manager.py:162: in cog_data_path
    base_data_path = Path(_base_data_path())
.tox/py38/lib/python3.8/site-packages/redbot/core/data_manager.py:134: in _base_data_path
    raise RuntimeError("You must load the basic config before you can get the base data path.")
E   RuntimeError: You must load the basic config before you can get the base data path.

The above exception was the direct cause of the following exception:
tests/cogs/test_audio.py:2: in <module>
    from redbot.pytest.audio import *
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/py38/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
.tox/py38/lib/python3.8/site-packages/redbot/pytest/audio.py:3: in <module>
    from redbot.cogs.audio import Audio
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/py38/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
.tox/py38/lib/python3.8/site-packages/redbot/cogs/audio/__init__.py:3: in <module>
    from .core import Audio
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/py38/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
.tox/py38/lib/python3.8/site-packages/redbot/cogs/audio/core/__init__.py:25: in <module>
    from . import abc, cog_utils, commands, events, tasks, utilities
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/py38/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
.tox/py38/lib/python3.8/site-packages/redbot/cogs/audio/core/tasks/__init__.py:4: in <module>
    from .lavalink import LavalinkTasks
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/py38/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
.tox/py38/lib/python3.8/site-packages/redbot/cogs/audio/core/tasks/lavalink.py:9: in <module>
    from ...manager import ServerManager
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/py38/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
.tox/py38/lib/python3.8/site-packages/redbot/cogs/audio/manager.py:60: in <module>
    LAVALINK_DOWNLOAD_DIR: Final[pathlib.Path] = data_manager.cog_data_path(raw_name="Audio")
.tox/py38/lib/python3.8/site-packages/redbot/core/data_manager.py:164: in cog_data_path
    raise RuntimeError(
E   RuntimeError: You must load the basic config before you can get the cog data path.
========================================================== short test summary info ===========================================================
ERROR tests/cogs/test_audio.py - RuntimeError: You must load the basic config before you can get the cog data path.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================== 1 error in 1.67s ==============================================================
ERROR: InvocationError for command /home/user/dev/Red-DiscordBot/.tox/py38/bin/pytest (exited with code 2)
```

jefetienne commented:
> I am of the opinion that this should be handled in the audio backend and not in our code, but I will wait to see what others have to say about this.

Oh, for lavalink?

aikaterna commented:
> > I am of the opinion that this should be handled in the audio backend and not in our code, but I will wait to see what others have to say about this.
> 
> Oh, for lavalink?

My opinion for this concept is that it's better handled inside Lavaplayer's bandcamp source manager, so that there is no webscraping code involved in the code for audio, and that it's contained/managed at the source.

aikaterna commented:
https://github.com/Walkyst/lavaplayer-fork/blob/original/main/src/main/java/com/sedmelluq/discord/lavaplayer/source/bandcamp/BandcampAudioSourceManager.java if you're interested in leaving a PR on Walkyst's lavaplayer fork, which is what we use in our Lavalink.jar files.

jefetienne commented:
> > > I am of the opinion that this should be handled in the audio backend and not in our code, but I will wait to see what others have to say about this.
> > 
> > 
> > Oh, for lavalink?
> 
> My opinion for this concept is that it's better handled inside Lavaplayer's bandcamp source manager, so that there is no webscraping code involved in the code for audio, and that it's contained/managed at the source.

I can definitely see the logic in that and would have no problem bringing it up to Lavaplayer.

I might be wrong on this, but I think even if that is done, there would still need to be some changes on Red's end. For example, on `redbot/cogs/audio/core/utilities/validation.py`, we have a validation method running at the beginning of `player.py`'s `PlayerCommands.command_play(..)` that would reject the custom domain URL that I think would occur even before Lavaplayer would parse it themselves:

```py
def is_url_allowed(self, url: str) -> bool:
        valid_tld = [
            "youtube.com",
            "youtu.be",
            "soundcloud.com",
            "bandcamp.com",
            "vimeo.com",
            "twitch.tv",
            "spotify.com",
            "localtracks",
        ]
        query_url = urlparse(url)
        url_domain = ".".join(query_url.netloc.split(".")[-2:])
        if not query_url.netloc:
            url_domain = ".".join(query_url.path.split("/")[0].split(".")[-2:])
        return True if url_domain in valid_tld else False
```

aikaterna commented:
> that would reject the custom domain URL that I think would occur even before Lavaplayer would parse it themselves:

You can use arbitrary urls once [p]audioset restrict is toggled off. This is on by default to protect people from making queries to domains that are not major services. Yes, custom bandcamp urls are redirects to bandcamp pages, but they are an external url that would fall under this classification.

jefetienne commented:
> > that would reject the custom domain URL that I think would occur even before Lavaplayer would parse it themselves:
> 
> You can use arbitrary urls once [p]audioset restrict is toggled off. This is on by default to protect people from making queries to domains that are not major services. Yes, custom bandcamp urls are redirects to bandcamp pages, but they are an external url that would fall under this classification.

Interesting, I wasn't aware of the setting. I tested it now on the default V3/develop branch and it looks like Lavaplayer still has an issue trying to play the songs (my bot returned a "[Nothing Found] Unkown file format.")

I can close this PR then if there isn't anything else to be said and look into working on Lavaplayer, although I'm not sure how many extra steps it would take for me to test from Lavaplayer -> Lavalink -> Red.

aikaterna commented:
> my bot returned a "[Nothing Found] Unkown file format."

Correct - lavaplayer itself does not currently resolve custom Bandcamp urls to the playable urls, which is where the main issue is for this feature request - lavaplayer should parse these and provide a track object to Lavalink, which Red could then play. That's the main reason I am trying to say this functionality should be in lavaplayer itself and not at the Red level.

aikaterna commented:
And, well, I guess I should add that I am also against the idea of maintaining webscrapers inside Red because then we would always be responsible for updating it when the page parsing changes. Right now all of the webscraping involved in this entire process is confined to lavaplayer source managers.

jefetienne commented:
Sounds good, closing now.

aikaterna commented:
Thank you for the attempt though, imo this is something that should be addressed at some point but not on the Red side of things.
